src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,373 @@
+            if (callerDeclarationNode == null)
+            {
+                return;
+            }
+
+            var callerSymbol = semanticModel.GetDeclaredSymbol(callerDeclarationNode, cancellationToken)
+                ?? semanticModel.GetSymbolInfo(callerDeclarationNode, cancellationToken).GetAnySymbol();
+            if (callerSymbol == null)
+            {
+                return;
+            }
+
+            var containsAwaitExpression = ContainsAwaitExpression(inlineExpression, calleeMethodNode);
+            if (containsAwaitExpression)
+            {
+                // If there is nested 'await', check if the caller is a method, and its return type
+                // is void or AwaitableType because after move the 'await' to the caller it must be async
+                // Example that shouldn't offer refactoring:
+                // int Caller()
+                // {
+                //    var x = Callee();
+                //    return 1;
+                // }
+                // async Task Callee() => await Task.Delay(await Task.FromResult(100));
+                // Example that should offer refactoring (because it is safe to change the caller to async):
+                // void Caller()
+                // {
+                //    var x = Callee();
+                // }
+                // async Task Callee() => await Task.Delay(await Task.FromResult(100));
+                var canInlineAwaitExpression = callerSymbol is IMethodSymbol callerMethodSymbol
+                       && (callerMethodSymbol.IsAsync ||
+                           callerMethodSymbol.ReturnsVoid ||
+                           calleeMethodSymbol.IsAwaitableNonDynamic(semanticModel, callerDeclarationNode.SpanStart));
+                if (!canInlineAwaitExpression)
+                {
+                    return;
+                }
+            }
+
+            var statementContainsCallee = calleeMethodInvocationNode.GetAncestor<TStatementSyntax>();
+            if (statementContainsCallee == null)
+            {
+                return;
+            }
+
+            var invocationOperation = semanticModel.GetOperation(calleeMethodInvocationNode, cancellationToken) as IInvocationOperation;
+            if (invocationOperation == null)
+            {
+                return;
+            }
+
+            var codeActions = GenerateCodeActions(
+                document,
+                calleeMethodInvocationNode,
+                calleeMethodSymbol,
+                calleeMethodNode,
+                inlineExpression,
+                callerDeclarationNode,
+                callerSymbol,
+                containsAwaitExpression,
+                statementContainsCallee,
+                invocationOperation);
+
+            var nestedCodeAction = new MyNestedCodeAction(
+                string.Format(FeaturesResources.Inline_0, calleeMethodSymbol.ToNameDisplayString()),
+                codeActions,
+                isInlinable: false);
```

Comment:
preference is to make this inlinable and do the following:

`
Inline method
  Inline '{0}'
  Inline and keep '{0}'
`

markthis as `isInlinable: true` so taht we can elide the parent if there aren't many code actions visible.